### ISSUE TITLE
error: You must be logged in to the server (Unauthorized)

### ISSUE THREAD

**Issue by tomweston** [heptio/authenticator#92]
## EKS Cluster Authentication

**TLDR;** Not sure what I'm missing here 😢 ??

```
error: You must be logged in to the server (Unauthorized)
```
## Steps Taken:

- **1. EKS Cluster created in AWS Console**

- **2. IAM Role created with the following steps:**

[https://github.com/heptio/authenticator#1-create-an-iam-role](https://github.com/heptio/authenticator#1-create-an-iam-role)

- **3. STS Assume role allowed on Role from user account in AWS_PROFILE**

```
{
    "Sid": "KubernetesAdmin",
    "Effect": "Allow",
    "Action": "sts:AssumeRole",
    "Resource": "arn:aws:iam::012345678910:role/KubernetesAdmin"
}
```
- **4. Values from UI copied and placed into a new config ~/.kube/config-clustername as per AWS Documentation:**

[https://docs.aws.amazon.com/eks/latest/userguide/create-kubeconfig.html](https://docs.aws.amazon.com/eks/latest/userguide/create-kubeconfig.html)

```
apiVersion: v1
clusters:
- cluster:
    server: <APIServerEndpointURL>
    certificate-authority-data: <CAbase64blob>
  name: kubernetes
contexts:
- context:
    cluster: kubernetes
    user: aws
  name: aws
current-context: aws
kind: Config
preferences: {}
users:
- name: aws
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1alpha1
      command: heptio-authenticator-aws
      args:
        - "token"
        - "-i"
        - "clustername"
        - "-r"
        - "arn:aws:iam::012345678910:role/KubernetesAdmin"
```

- **5. Kubectl get nodes using new config:**

```
$ kubectl --kubeconfig ~/.kube/config-clustername get nodes
error: You must be logged in to the server (Unauthorized)

```
😢 What's missing?

**Comment by mattlandis** [comment]
The role/user that created the cluster through the console will have admin (system:masters) permissions on the cluster.  To add another user/role you need to edit the aws-auth configmap in the kube-system namespace. (If it doesn't exist the eks docs cover adding it as part of adding nodes).

**Comment by thedevelopnik** [comment]
Yeah same problem, logs in the pod say that there's a bad certificate error, so it seems like heptio-authenticator-aws isn't providing a certificate that the stuff on the server that it generated likes. I'm using KOPS, k8s 1.9. Followed the Kops instructions, including the rolling-update. I did vary from the official instructions by using an existing role, which is an account specific role, not cross-account. The account that it is for is the same one that the cluster is running in.

**Comment by christopherhein** [comment]
@thedevelopnik Did you make sure to change the `volumeMounts` to `/srv` as step 4. says… most people I've talked to have missed that at least that are running into either permission issues.

> Update the Authenticator DaemonSet's state and output volumes to both use /srv/kubernetes/heptio-authenticator-aws/ for their hostPaths.

**Comment by christopherhein** [comment]
@tomweston wherein the EKS setup are you? Have you deployed the node `ConfigMap` and are you trying to add additional roles to have authentication or are you trying to deploy the Node `ConfigMap` with this new role??

**Comment by mickengland** [comment]
We are getting the same error from a fresh cluster created with a user without MFA after initial problems with that. We are working through this guide https://docs.aws.amazon.com/eks/latest/userguide/getting-started.html and at the point of verifying the client configuration with this:

$ kubectl get svc
error: the server doesn't have a resource type "svc"

If I then do kubectl get serviceAccount i get this:

$ kubectl get serviceAccount
error: You must be logged in to the server (Unauthorized)

kubectl apply -f aws-auth-cm.yaml

Gives me this:
error: You must be logged in to the server (the server has asked for the client to provide credentials)

So we are at the point where we have the control plain running and nodes created with the Cloud Formation stack but can't join them to the cluster due to these authentication issues.

We have confirmed that the aws credentials work because we can run this:

$ aws eks describe-cluster --name staging-eks --query cluster.status
"ACTIVE"

**Comment by nckturner** [comment]
@mickengland Did you double check the endpoint, CA, and cluster name in your kube config?

**Comment by mattlandis** [comment]
When you do `aws sts get-caller-identity` the IAM identity needs to match the one you used to create the cluster.  Also the cluster-name in the kubeconfig exec needs to match the cluster's name in describe-cluster.

**Comment by mickengland** [comment]
@nckturner All of the details for config were retrieved directly from the cluster with the aws cli and pasted into the ~/kube/config file. Also the aws sts get-caller-identity matches the AIM identity used to create the cluster.

**Comment by tomweston** [comment]
**Thanks everyone for your feedback 😀**

So it appears the problem came down to the IAM identity that was used to create the cluster as pointed out by @mattlandis 

The trial cluster was created manually via the console using a cross role account eg. `developer @ myorg_account_a` 

The `developer` role has ALL permisions on EKS resources in `myorg_account_a` as well as `myorg_account_b` and `myorg_account_c` etc.

I wrongly assumed that the identity would match the one used when I had assumed the role to create the cluster in `myorg_account_a`.

### Steps to fix my mistake

- **1. Create a user account (in my case `eks` ) with permissions to access EKS in the intended account**

- **2. Gather the new CA and endpoint values and dump them into a new config. Im my case ~/.kube/config-testcluster**

- **3. Switch to this config: `export KUBECONFIG=$KUBECONFIG:~/.kube/config-testcluster`**

- **4. Add the new user to `~/.aws/credentials` and `~/.aws/config` then `export AWS_PROFILE=eks`**

- **5. Cross your fingers 🤞**

```
$ kubectl get svc
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes   ClusterIP   172.20.0.1   <none>        443/TCP   9m
```
**Bingo..!👌**

In our case, all of our K8s clusters are created and managed by terraform via CI using local account API Keys. So this would not have been a problem going forward. However, i'ts nice to bump into these Cross Account Role limitations with EKS now rather than later.

**Comment by mickengland** [comment]
AWS support resolved my issue for me.

Comment the lines below on your Kubeconfig:

# - "-r"
# - "arn:aws:iam::...

You're creating the cluster with user "user/eks-manager" and trying to access it with the role "MyRoleForAmazonEKS". 

This solved my problem.

**Comment by nckturner** [comment]
Awesome 👍 I think this can be closed.
